File: c/0001_TwoSum.c
// Problem: Two Sum (https://leetcode.com/problems/two-sum/) 
// Difficulty: Easy
// Date: 2021-11-23

/*
Given an array of integers nums and an integer target, return indices of the two
numbers such that they add up to target. You may assume that each input would
have exactly one solution, and you may not use the same element twice. You can
return the answer in any order.
*/

int* twoSum(int* nums, int numsSize, int target, int* returnSize){
    for (int i = numsSize - 1; i > 0; i--) {
        int required = target - nums[i];
        for (int j = i - 1; j >= 0; j--) {
            if (nums[j] == required) {
                int* result = malloc(2 * sizeof(int));
                if (result == NULL) {
                    return NULL;
                }
                
                result[0] = i;
                result[1] = j;
                *returnSize = 2;
                return result;
            }
        }
    }
    *returnSize = 0;
    return NULL;
}

File: py/0001_TwoSum.py
# 1. Two Sum

from typing import List


class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        nums = list(enumerate(nums))  # retain original indices
        nums.sort(key=lambda x: x[1])
        i, j = 0, len(nums) - 1
        while i <= j:
            required = target - nums[i][1]
            if required == nums[j][1]:
                return [nums[i][0], nums[j][0]]
            if required < nums[j][1]:
                j -= 1
            else:
                i += 1


"""
Runtime
70
ms
Beats
42.51%
of users with Python3
Memory
17.95
MB
Beats
9.47%
of users with Python3
5
"""
